File: tests/integration/stress_code_health_review.py
#!/usr/bin/env python3
"""
Stress test code_health_review against a built cs-mcp binary.

The goal is to catch intermittent embedded `cs` CLI failures, especially
the telemetry flush race that can surface as:

  NoSuchFileException: .../codescene-cli.log.jsonl
"""

import argparse
import os
import sys
import time
from pathlib import Path

sys.path.insert(0, str(Path(__file__).parent))

from fixtures import get_sample_files
from test_utils import MCPClient, create_git_repo, create_test_environment, extract_result_text, safe_temp_directory

TELEMETRY_RACE_MARKERS = ("NoSuchFileException", "codescene-cli.log.jsonl")


def _contains_telemetry_race_error(text: str) -> bool:
    return all(marker in text for marker in TELEMETRY_RACE_MARKERS)


def _build_env() -> dict[str, str]:
    env = create_test_environment()
    env.setdefault("CS_DISABLE_VERSION_CHECK", "1")
    env.pop("CS_DISABLE_TRACKING", None)
    return env


def _collect_new_stderr(client: MCPClient, cursor: int) -> tuple[str, int]:
    lines = client.stderr_lines[cursor:]
    return "\n".join(lines), len(client.stderr_lines)


def run_stress_test(executable: Path, iterations: int, timeout: int) -> int:
    with safe_temp_directory(prefix="cs_mcp_review_stress_") as test_dir:
        repo_dir = create_git_repo(test_dir, get_sample_files())
        review_target = repo_dir / "src/services/order_processor.py"

        env = _build_env()
        command = [str(executable)]

        client = MCPClient(command, env=env, cwd=str(repo_dir))
        if not client.start():
            print("FAIL: could not start MCP server")
            stderr = client.get_stderr()
            if stderr:
                print("Recent stderr:\n" + stderr)
            return 1

        init_response = client.initialize()
        if "result" not in init_response:
            print("FAIL: initialize did not return a result")
            print(init_response)
            client.stop()
            return 1

        print(f"Running {iterations} code_health_review calls...")

        start = time.time()
        stderr_cursor = 0
        total_failures = 0
        telemetry_failures = 0

        try:
            for i in range(1, iterations + 1):
                response = client.call_tool(
                    "code_health_review",
                    {"file_path": str(review_target)},
                    timeout=timeout,
                )

                response_text = extract_result_text(response)
                new_stderr, stderr_cursor = _collect_new_stderr(client, stderr_cursor)

                failed = False
                if "error" in response:
                    failed = True
                elif not response_text:
                    failed = True

                telemetry_error = _contains_telemetry_race_error(response_text) or _contains_telemetry_race_error(
                    new_stderr
                )
                if telemetry_error:
                    telemetry_failures += 1
                    failed = True

                if failed:
                    total_failures += 1
                    print(f"[{i}/{iterations}] FAIL")
                    if telemetry_error:
                        print("  telemetry race marker detected")
                    if "error" in response:
                        print(f"  response error: {response['error']}")
                    if response_text:
                        preview = response_text[:300].replace("\n", " ")
                        print(f"  response preview: {preview}")
                    if new_stderr:
                        print("  new stderr lines:")
                        print("  " + "\n  ".join(new_stderr.splitlines()[-8:]))
                elif i % 25 == 0 or i == iterations:
                    print(f"[{i}/{iterations}] OK")
        finally:
            client.stop()

        elapsed = time.time() - start
        print("\n--- Stress Test Summary ---")
        print(f"iterations:         {iterations}")
        print(f"total failures:     {total_failures}")
        print(f"telemetry failures: {telemetry_failures}")
        print(f"elapsed seconds:    {elapsed:.1f}")

        if total_failures == 0:
            print("PASS: no failures observed")
            return 0

        if telemetry_failures > 0:
            print("FAIL: telemetry race still observed")
        else:
            print("FAIL: non-telemetry errors observed")
        return 1


def main() -> int:
    parser = argparse.ArgumentParser(description="Stress test code_health_review for CLI stability")
    parser.add_argument("executable", type=Path, help="Path to cs-mcp binary")
    parser.add_argument(
        "--iterations",
        type=int,
        default=250,
        help="Number of code_health_review calls (default: 250)",
    )
    parser.add_argument(
        "--timeout",
        type=int,
        default=90,
        help="Per-call timeout in seconds (default: 90)",
    )
    args = parser.parse_args()

    if not args.executable.exists():
        print(f"Error: executable not found: {args.executable}")
        return 1

    if not os.getenv("CS_ACCESS_TOKEN"):
        print("Error: CS_ACCESS_TOKEN is not set")
        return 1

    if args.iterations < 1:
        print("Error: --iterations must be >= 1")
        return 1

    return run_stress_test(args.executable, args.iterations, args.timeout)


if __name__ == "__main__":
    sys.exit(main())

File: tests/e2e/tests/stress_code_health_review.rs
//! Stress test for `code_health_review` under repeated invocations.
//!
//! Validates that a single MCP server instance handles many sequential
//! `code_health_review` calls without telemetry race conditions or failures.
//! Telemetry tracking is deliberately enabled to exercise the file-locking path.

use super::*;

const DEFAULT_ITERATIONS: usize = 50;
const TIMEOUT_SECS: u64 = 90;
const TELEMETRY_RACE_MARKERS: &[&str] = &["NoSuchFileException", "codescene-cli.log.jsonl"];

fn contains_telemetry_race_error(text: &str) -> bool {
    TELEMETRY_RACE_MARKERS.iter().all(|marker| text.contains(marker))
}

fn check_single_call(client: &mut MCPClient, file_path: &Path) -> (bool, bool) {
    let response = client.call_tool(
        "code_health_review",
        json!({"file_path": file_path.to_string_lossy()}),
        Duration::from_secs(TIMEOUT_SECS),
    );

    match &response {
        Ok(resp) => {
            let text = extract_result_text(resp);
            let has_race = contains_telemetry_race_error(&text);
            let failed = text.is_empty() || has_race;
            (failed, has_race)
        }
        Err(_) => (true, false),
    }
}

fn run_iterations(client: &mut MCPClient, file_path: &Path) -> (usize, usize) {
    let mut total_failures = 0;
    let mut telemetry_failures = 0;

    for i in 1..=DEFAULT_ITERATIONS {
        let (failed, has_race) = check_single_call(client, file_path);
        if has_race {
            telemetry_failures += 1;
        }
        if failed {
            total_failures += 1;
        }
        if i % 10 == 0 || i == DEFAULT_ITERATIONS {
            eprintln!("[{i}/{DEFAULT_ITERATIONS}] failures so far: {total_failures}");
        }
    }

    let stderr = client.get_stderr();
    if contains_telemetry_race_error(&stderr) {
        eprintln!("Telemetry race markers found in server stderr");
        telemetry_failures += 1;
        total_failures += 1;
    }

    (total_failures, telemetry_failures)
}

pub fn test_stress_code_health_review() {
    let executable = find_or_build_executable();
    let backend = create_backend(executable);
    let temp_dir = create_temp_dir("cs_mcp_stress_").expect("temp dir");
    let repo_dir = create_git_repo(temp_dir.path(), &get_sample_files()).expect("git repo");

    let mut base = base_env();
    base.retain(|k, _| k != "CS_DISABLE_TRACKING");

    let env = backend.get_env(&base, &repo_dir);
    let env_vec: Vec<(String, String)> = env.into_iter().collect();
    let command = backend.get_command(&repo_dir);

    let mut client = make_client(&command, &env_vec, &repo_dir);
    assert!(client.start(), "Server should start");
    client.initialize().expect("Initialize should succeed");

    let review_target = repo_dir.join("src/services/order_processor.py");
    let (total_failures, telemetry_failures) =
        run_iterations(&mut client, &review_target);

    assert_eq!(
        total_failures, 0,
        "Stress test had {total_failures} failures ({telemetry_failures} telemetry)"
    );
}
